analysis/topology/banach_contraction.lean

```diff
@@ -0,0 +1,284 @@
+/-
+Copyright (c) 2018 Rohan Mitta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rohan Mitta, Kevin Buzzard, Alistair Tucker
+-/
+--Working towards a proof of the Banach Contraction Theorem
+import analysis.limits
+import analysis.topology.metric_sequences
+import data.complex.exponential
+
+open filter
+lemma tendsto_succ (X : Type*) (f : ℕ → X) (F : filter X) (H : tendsto f at_top F) :
+tendsto (λ n, f (n + 1)) at_top F :=
+tendsto.comp (tendsto_def.2 $ λ U HU,
+  let ⟨a,Ha⟩ := mem_at_top_sets.1 HU in
+  mem_at_top_sets.2 ⟨a,λ x Hx,Ha _ $ le_trans Hx $ by simp⟩) H
+
+local attribute [instance, priority 0] classical.prop_decidable
+noncomputable theory
+
+open function
+--The following material comes from "Metric Spaces and Topology" by Sutherland
+--Half of Proposition 17.4
+theorem complete_of_complete_of_uniform_cts_bij {α : Type*} [metric_space α] {β : Type*} [metric_space β] (f : α → β)
+    (g : β → α) (Hf : uniform_continuous f) (Hg : uniform_continuous g) (left_inv : function.left_inverse g f)
+    (right_inv : function.right_inverse g f) : complete_space α → complete_space β :=
+begin
+  rintro ⟨H1⟩,
+  split,
+  intros filt Hfilt,
+  cases H1 (cauchy_map Hg Hfilt) with x H_converges_to_x,
+  existsi f x,
+  rw [map_le_iff_le_comap,
+      ←filter.map_eq_comap_of_inverse (id_of_right_inverse right_inv) (id_of_left_inverse left_inv)] at H_converges_to_x,
+  exact le_trans H_converges_to_x (continuous.tendsto Hf.continuous x)
+end
+
+--Proposition 17.4
+theorem complete_iff_of_uniform_cts_bij {α : Type*} [metric_space α] {β : Type*} [metric_space β] (f : α → β) 
+    (g : β → α) (Hf : uniform_continuous f) (Hg : uniform_continuous g) (left_inv : function.left_inverse g f)
+    (right_inv : function.right_inverse g f) : complete_space α ↔ complete_space β := 
+        ⟨complete_of_complete_of_uniform_cts_bij f g Hf Hg left_inv right_inv,
+        complete_of_complete_of_uniform_cts_bij g f Hg Hf right_inv left_inv⟩
+
+open nat
+def iteration_map {α : Type*} (f : α → α) (start : α) : ℕ → α
+| zero := start
+| (succ x) := f (iteration_map x)
+
+--Definition 17.24
+def is_contraction {α : Type*} [metric_space α] (f : α → α) := 
+∃ (k : ℝ) (H1 : k < 1) (H2 : 0 < k), ∀ (x y : α), dist (f x) (f y) ≤ k* (dist x y)

```

Comment:
maybe put some more space in, I was wondering what the `k*` function is

Reply:
I'm not sure this is yet ready to review. Between Rohan Mitta and myself we have a lot of redundancy. There is a completely parallel proof of the main theorem here https://github.com/agjftucker/mathlib/blob/Banach/analysis/topology/banach_contraction.lean

Reply:
I should have said: it would be great to have your feedback :)

Reply:
ok, remove the [WIP] when ready